lib/pbench/agent/redis.py

```diff
@@ -222,39 +222,77 @@ def emit(self, record):
             self.counter += 1
 
 
-def wait_for_conn_and_key(redis_server, key, prog, redis_host, redis_port):
+def wait_for_conn_and_key(redis_server: redis.Redis, key: str, prog: str) -> str:
     """wait_for_conn_and_key - convenience method of both the Tool Meister and
     the Tool Data Sink to startup and wait for an initial connection to the
     Redis server, and for the expected key to show up.
+
+    Arguments:
+
+        redis_server: a Redis client object
+        key: the key name as a string
+        prog: the program name of our caller
+
+    Returns the payload value of the key as a string.
+
+    If successful on the first connection attempt, and the value of the key is
+    retrieved, no messages are issued on stdout.  If the key does not exist yet,
+    we'll report the missing key every 10 seconds from the last successful
+    connection.
+
+    If on the first connection attempt the connection to the Redis server fails,
+    we emit a message to that effect, once.  After that, we'll report each time
+    we are disconnected or re-connected to the Redis server.
     """
     # Loop waiting for the key to show up.
     connected = None
     payload = None
+    attempts = 0
+    errors = 0
+    missing = 0
     while payload is None:
+        attempts += 1
         try:
             payload = redis_server.get(key)
         except redis.ConnectionError:
+            errors += 1
+            # Reset the missing count if we get disconnected
+            missing = 0
             if connected is None:
+                # Only emit this message once for the initial attempt.
                 print(
-                    f"{prog}: waiting to connect to redis server {redis_host}:{redis_port}",
+                    f"{prog}: waiting to connect to Redis server {redis_server}",
                     flush=True,
                 )
                 connected = False
             elif connected:
+                # We always report disconnections.
                 print(
-                    f"{prog}: disconnected from redis server {redis_host}:{redis_port}",
+                    f"{prog}: disconnected from Redis server {redis_server}",
                     flush=True,
                 )
                 connected = False
             time.sleep(1)
```

Comment:
Do you really want to maintain a separate delay for the non-connected retry (here) and the connected retry (at line 288)?  I can imagine how that could be useful, but, since they are both hard-coded to use the same interval, I would suggest combining them simply by using a single, common call at the bottom of the loop.

Reply:
Related: it's assumed in the code that 10 delays equals one second (the message says so), so rather than an anonymous inline `1` we'd be better off with a constant with a meaningful name like `ONE_SECOND` to make the semantic assumption clear.

Reply:
> Related: it's assumed in the code that 10 delays equals one second (the message says so)

10 delays equals one second?  Can you help me see where the message says that?

>, so rather than an anonymous inline `1` we'd be better off with a constant with a meaningful name like `ONE_SECOND` to make the semantic assumption clear.

If we were sleeping for a tenth of a second, I could see making a constant, ... perhaps, but I am not sure the all the mechanism to do that is worth the additional clarity.